File: ups/setups.pl
package ups;

BEGIN {
    my ($setups_sh, $c1);

    $setups_sh = "${Setups_Home}/setups.sh";
    $::ENV{UPS_SHELL} = "sh";
    $c1=` . $setups_sh; 
	  echo "\$::ENV{UPS_DIR}=q(\$UPS_DIR); \$::ENV{PRODUCTS}=q(\$PRODUCTS); \$::ENV{SETUP_UPS}=q(\$SETUP_UPS);"`;

    unless (eval $c1) {
	die("setups.pl failed $@");
    };
}


sub setup {
 
    my ($file, $c1);

    $file = `$::ENV{UPS_DIR}/bin/ups setup $_[0]`;


    chomp($file);

    open( SETUP, ">>$file" );

    print SETUP <<'EOF';

echo "--------------cut here-------------"
perl -e 'for $v (keys(%::ENV)) { 
	    ($fix= $::ENV{$v}) =~ s/[()]/\$&/go;
	    print "\$::ENV{$v} = q($fix);\n" 
	 }'

EOF
    close(SETUP);

    $c1 = `/bin/sh $file`;
    $c1 =~ s/.*--------------cut here-------------//s;

    %::ENV = ();
    unless (eval $c1) {
	die("setup failed $@");
    };
}
 
sub unsetup {
 
    my ($file, $c1);

    $file = `$::ENV{UPS_DIR}/bin/ups unsetup $_[0]`;


    chomp($file);

    open( SETUP, ">>$file" );

    print SETUP <<'EOF';

echo "--------------cut here-------------"
perl -e 'for $v (keys(%::ENV)) { 
	    ($fix= $::ENV{$v}) =~ s/[()]/\$&/go;
	    print "\$::ENV{$v} = q($fix);\n" 
	 }'

EOF

    close(SETUP);

    $c1 = `/bin/sh $file`;
    $c1 =~ s/.*--------------cut here-------------//s;

    %::ENV = ();
    unless (eval $c1) {
	die("unsetup failed $@");
    };
}

sub use_perl {
    my ($v) = $_[0];
    if ( $v eq "" ) {
	$vm = '.*';
    } else {
	$vm = $v;
    }
    if ( ! defined($::ENV{SETUP_PERL}) || $::ENV{SETUP_PERL} !~ m/perl $vm / ) {
	setup("perl $v");
	exec "$::ENV{PERL_DIR}/bin/perl", "-wS", $::0, @::ARGV;
	die( "cannot exec $::ENV{PERL_DIR}/bin/perl" );
    }
}
 
1;

File: ups/setups.py
import os
import re
import sys


class setups:
    def __init__(self):
	os.environ['UPS_SHELL'] = "sh"
	f = os.popen( '. ${Setups_Home}/setups.sh; echo os.environ\\[\\"UPS_DIR\\"\\]=\\"$UPS_DIR\\"; echo os.environ\\[\\"PRODUCTS\\"\\]=\\"$PRODUCTS\\"; echo os.environ\\[\\"SETUP_UPS\\"\\]=\\"$SETUP_UPS\\"')
	exec f.read()
	f.close()

    def setup(self, arg):
	self.inhaleresults(os.environ["UPS_DIR"] + '/bin/ups setup ' + arg)

    def unsetup(self, arg):
	self.inhaleresults(os.environ["UPS_DIR"] + '/bin/ups unsetup ' + arg)

    def inhaleresults(self, cmd):

	f = os.popen(cmd)
	filename = f.read()
	filename = filename[0:-1]
	f.close()

	setup = open(filename,"a")
	setup.write(
	    ''' 
echo "--------------cut here-------------"
cat <<'EOF' | python
import os
import re
for v in os.environ.keys():
    fix= re.sub( "\'", "\\'", os.environ[v])
    print "os.environ['"+v+"'] = '" + fix + "'" 
EOF
	    ''')
	setup.close()

	f = os.popen("/bin/sh " + filename)
	c1 = f.read()
	f.close()
	c1 = re.sub( '.*--------------cut here-------------', '', c1)
	os.environ = {}
	exec c1

    def use_python(self,v):

        self.use_package("python", v, "SETUP_PYTHON")

    def use_package(self, p, v, s):
	if ( v == "" ):
	    vm = '.*';
	else:
	    vm = v;

	if ( not os.environ.has_key(s) or None == re.search(p+' '+vm,os.environ.get(s,''))):
	    if os.environ.has_key(s):
		self.unsetup(p)
	    self.setup(p + ' ' + v)
	    sys.argv.insert(0, os.environ['PYTHON_DIR']+'/bin/python')
	    os.execve( os.environ['PYTHON_DIR']+'/bin/python', sys.argv, os.environ )

    def ups(self, *args):
        f = os.popen( os.environ['UPS_DIR'] + '/bin/ups ' + string.join(args))
        res = f.read()
        f.close()
        return res
